File: src/main/java/com/telegram/codex/integration/telegram/application/webhook/UnsupportedMessageHandler.java
package com.telegram.codex.integration.telegram.application.webhook;

import com.telegram.codex.conversation.domain.MessageConstants;
import com.telegram.codex.integration.telegram.application.port.out.TelegramGateway;
import com.telegram.codex.integration.telegram.domain.InboundMessage;
import com.telegram.codex.integration.telegram.domain.webhook.TelegramMessage;
import com.telegram.codex.integration.telegram.domain.webhook.TelegramUpdate;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class UnsupportedMessageHandler {

    private final TelegramGateway telegramClient;

    public UnsupportedMessageHandler(TelegramGateway telegramClient) {
        this.telegramClient = telegramClient;
    }

    public boolean handle(InboundMessage message, TelegramUpdate update) {
        if (message != null && !message.unsupported()) {
            return false;
        }
        if (update == null) {
            return true;
        }
        String chatId = extractChatId(update);
        if (chatId != null) {
            telegramClient.sendMessage(chatId, MessageConstants.UNSUPPORTED_MESSAGE, List.of(), false);
        }
        return true;
    }

    private String extractChatId(TelegramUpdate update) {
        TelegramMessage message = update.message();
        if (message == null || message.chat() == null || message.chat().id() == null) {
            return null;
        }
        return String.valueOf(message.chat().id());
    }
}

File: src/telegram/inbound/unsupported_message_handler.rs
use std::sync::Arc;

use anyhow::Result;

use crate::telegram::shared::inbound_message::InboundMessage;
use crate::telegram::shared::message_constants;
use crate::telegram::shared::telegram_types::{TelegramGateway, TelegramUpdate};

pub struct UnsupportedMessageHandler {
    telegram_client: Arc<dyn TelegramGateway>,
}

impl UnsupportedMessageHandler {
    pub fn new(telegram_client: Arc<dyn TelegramGateway>) -> Self {
        Self { telegram_client }
    }

    /// Returns `true` once the "I cannot handle this" notice has been sent, or
    /// when there was no chat to reply to at all.
    pub async fn handle(&self, message: Option<&InboundMessage>, update: Option<&TelegramUpdate>) -> Result<bool> {
        if let Some(message) = message
            && !message.unsupported()
        {
            return Ok(false);
        }
        let chat_id = message.map(|message| message.chat_id.clone()).or_else(|| {
            update
                .and_then(|update| update.message.as_ref())
                .and_then(|message| message.chat.as_ref())
                .and_then(|chat| chat.id.as_ref())
                .map(|id| id.as_string())
        });
        if let Some(chat_id) = chat_id.filter(|chat_id| !chat_id.is_empty()) {
            self.telegram_client.send_message(&chat_id, Some(message_constants::UNSUPPORTED_MESSAGE), &[], false).await?;
        }
        Ok(true)
    }
}
